nixos/displayManager: introduce defaultSession

hedning opened the pull request:
###### Motivation for this change

Integrate `services.xserver.displayManager.sessionPackages` better with the existing session system.

resolves https://github.com/NixOS/nixpkgs/issues/54089

###### services.xserver.displayManager.defaultSession

There's two ways of providing graphical sessions now:
1. `displayManager.session` via. `desktopManager.session` and
  `windowManager.session`
2. `displayManager.sessionPackages`

`sessionPackages` doesn't make a distinction between desktop and window
managers. This makes selecting a session provided by a package using
`desktopManager.default` nonsensical.

We therefor introduce `displayManager.defaultSession` which can select a session
from either `displayManager.session` or `displayManager.sessionPackages`.

It will default to `desktopManager.default + windowManager.default` as before.
If the dm default is "none" it will select the first provided session from
`sessionPackages`.

###### services.xserver.displayManager.sessionPackages

Note `extraSessionFilePackages` is renamed to the less verbose `sessionPackages`.

When adding a package, `foo`, to `sessionPackages` session names are now required to be specified in `foo.passthru.providedSessions`, this is used to validate `defaultSession` at evaluation time.

At build time `mkDesktops` will check that the sessions provided by a package is actually present, bailing out if not, hopefully preventing typos etc.

closes #39871 

###### Things done

Ran these tests succesfully:
- [x] gnome3
- [x] sddm
- [x] plasma5
- [x] lightdm
- [x] xfce



------------------------

After `default` options deprecation:

- [x] gdm+xfce+openbox
- [x] lightdm+xfce+openbox
- [x] lightdm+gnome-flashback-metacity

hedning committed:
nixos/display-manager: extraSessionFilePackages -> sessionPackages

hedning committed:
nixos/sessionPackages: provide session names in passthru

We want access to the valid session names at evaluation time.

hedning committed:
nixos/displayManager: check for provided sessions in mkDesktops

hedning committed:
nixos/displayManager: introduce defaultSession

There's two ways of providing graphical sessions now:
- `displayManager.session` via. `desktopManager.session` and
  `windowManager.session`
- `displayManager.sessionPackages`

`sessionPackages` doesn't make a distinction between desktop and window
managers. This makes selecting a session provided by a package using
`desktopManager.default` nonsensical.

We therefor introduce `displayManager.defaultSession` which can select a session
from either `displayManager.session` or `displayManager.sessionPackages`.

It will default to `desktopManager.default + windowManager.default` as before.
If the dm default is "none" it will select the first provided session from
`sessionPackages`.

jtojnar commented:
Should we perhaps deprecate `desktopManager.default` and `windowManager.default`?

hedning commented:
Hmm, perhaps, but I'd imagine they're both used quite heavily. Also recreating the the `desktopManager.default+windowManager.default` manually also seems like chore, though I'm not certain how common it us to mix those in reality.

hedning commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -319,6 +319,38 @@ in
+        description = "Default graphical session (only effective for LightDM and SDDM).";
+        apply = d: let
+            sessionNames = cfg.displayManager.session.names ++
+              (concatMap (p: p.providedSessions) cfg.displayManager.sessionPackages);
```
I'm abusing `apply` here a bit to get a sane error message. Could move it to the type check, but would still need an explicit throw to get an error which actually tells the user what's wrong.

jtojnar replied:
Could we use `lib.assertMsg`?

hedning replied:
That would work.

hedning commented:
We should probably add some documentation too.

hedning commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -263,11 +268,24 @@ in
+              'passthru.providedSessions'. This is required when used as a session package.
+
+              The session names can be looked up in:
+                ${p}/share/xsessions
```
`assertMsg` worked well, thanks. Added a proper error message for `sessionPackages` which explains where and what things go.

hedning committed:
nixos/lightdm: enable wayland sessions

Note: can't launch gnome on wayland due to duplicate entry names:
  https://github.com/CanonicalLtd/lightdm/issues/16

jtojnar commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -263,11 +268,24 @@ in
+            (package.check p && p ? passthru.providedSessions
+            && p.providedSessions != [] && all isString p.providedSessions)
+            ''
+              Package, '${p.name}', didn't specificy any session names, as strings, in
```
Perhaps the message should only be asserted to the three last conjuncts – if `package.check p` fails, there might not be `p.name`.

jtojnar replied:
```suggestion
              Package, '${p.name}', did not specify any session names, as strings, in
```

hedning commented:
Added `share/wayland-sessions` to lightdm's `sessions-directory`, this won't work with gnome due to duplicate names, but is necessary when having a default session which is only found there. It should also be useful for sway et. al. Slim doesn't seem to support default session.

aanderse commented:
ping (triage)

worldofpeace commented:
Last time I checked @hedning was rethinking this approach.

hedning commented:
So it's possible to reuse `desktopManager.default` for selecting sessions from `sessionPackages`. But we still need to construct a `defaultSession` for lightdm/sddm which require the same logic as in `displayManager.defaultSession` (ie. checking if the name cannot be found in `displayManager.session.names` it's available in `displayManager.sessionPackages`).

So thinking on this some more I think it might be better to stick with an explicit `defaultSession` that the various display managers can use without trouble. The downside would be users who've set `desktopManager.default = "gnome"` which now would fail with an unhelpful error message. And somewhat a proliferation of similar options.

So I guess the question is this:
Do we want to select default sessions using `desktopManager.default` (and `windowManager.default`) or add a unified option that can override these?

jtojnar commented:
I would just remove `desktopManager.default` and `windowManager.default`, and generate xsession per `desktopManager.default` × `windowManager.default` pair. We already have d×w desktop files, and the xsession file are rather small and most people will not have many such pairs. We could even make `sessionPackage` per `displayManager.session`, which would simplify the modules further.

hedning commented:
> I would just remove desktopManager.default and windowManager.default

Yeah, if we're fine breaking some backwards compatibility, that would be ideal.

> We could even make sessionPackage per displayManager.session, which would simplify the modules further.

Not sure I follow, something like `services.xserver.displayManager.session.package`?

jtojnar commented:
> Yeah, if we're fine breaking some backwards compatibility, that would be ideal.

I am always up for that :smiling_imp: 

> Not sure I follow, something like `services.xserver.displayManager.session.package`?

We could remove the `apply` from `displayManager.session`. And add something like the following to the display manager module:

```nix
  services.xserver.displayManager.sessionPackages =
    let
      dms = filter (s: s.manage == "desktop") cfg.session;
      wms = filter (s: s.manage == "window") cfg.session;
    in
      map
        (d: w: let
          wmCommand = ...;
          script = ...;
        in writeTextFile {
          name = "${dmName}${optionalString (w.name != "none") ("+" + w.name)}-xsession";
          destination = "/share/xsessions/${dmName}${optionalString (w.name != "none") ("+" + w.name)}.desktop";
          text = ''
            [Desktop Entry]
            Name=${dmName}${optionalString (w.name != "none") ("+" + w.name)}
            Exec=${script}
            TryExec=${wmCommand}
            Type=Application
            DesktopNames=GNOME-Flashback;GNOME;
          '';
        } // {
          providedSessions = [ "${dmName}${optionalString (w.name != "none") ("+" + w.name)}" ];
        })
        (crossLists dms wms);
```

jtojnar commented:
I do think we would not even need to break any options, we can just deprecate the old options and assert that user is not setting both old and new default option.

hedning commented:
Ah, right, that makes sense. And deprecation sounds good.

jtojnar committed:
nixos/displayManager: deprecate separate options for default wm/dm

The upstream session files display managers use have no concept of sessions being composed from
desktop manager and window manager. To be able to set upstream session files as default
session, we need a single option. Having two different ways to set default session would be confusing,
though, so we decided to deprecate the old method.

We also created separate script for each session, just like we already had a separate desktop
file for each one, and started using displayManager.sessionPackages mechanism to make the
session handling more uniform.

jtojnar commented:
Guess I need to rebase for borg to build this with all the conflicts.

jtojnar commented:
@GrahamcOfBorg test gnome3 gnome3-xorg i3wm installer lightdm pantheon plasma5 sddm xautolock xfce xmonad xrdp xss-lock yabar

worldofpeace commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -117,93 +108,43 @@ let
-        Comment=
-        EODESKTOP
-      '') names}
+      mkdir -p "$out/share/"{xsessions,wayland-sessions}
```
The if below that does the `lndir` does this for `wayland-sessions` but not `xsessions`.
Can you drop this and add `mkdir` to the xsessions `if`?

worldofpeace commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -365,6 +373,60 @@ in
+        dms = filter (s: s.manage == "desktop") cfg.displayManager.session;
+        wms = filter (s: s.manage == "window") cfg.displayManager.session;
+
+        # file executed by the
```
I believe this is an incomplete comment

worldofpeace commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -297,16 +250,50 @@ in
+            else
+              throw "No desktop session packages installed";
+        example = "gnome";
+        description = "Default graphical session (only effective for LightDM and SDDM).";
```
This is no longer valid because we have the little script in gdm.

worldofpeace commented on nixos/modules/services/x11/window-managers/default.nix:
```diff
@@ -59,15 +59,14 @@ in
-          else
-            throw "Default window manager (${defaultWM}) not found.";
+        description = ''
+          Deprecated: Please use <xref linkend="opt-services.xserver.displayManager.defaultSession"/> instead.
```
Can the deprecated be bold?

worldofpeace commented on nixos/modules/services/x11/desktop-managers/default.nix:
```diff
@@ -86,23 +86,14 @@ in
-                ${concatMapStringsSep "\n  " (p: p.name) config.services.xserver.displayManager.extraSessionFilePackages}
-            '' defaultDM;
+        description = ''
+          Deprecated: Please use <xref linkend="opt-services.xserver.displayManager.defaultSession"/> instead.
```
can the `Deprecated` be bold?

worldofpeace commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -365,6 +373,63 @@ in
+              TryExec=${script}
+              Exec=${script}
+              Name=${sessionName}
+              Comment=
```
Would it be worth adding a `DesktopNames` key here?

jtojnar replied:
Hmm https://gitlab.freedesktop.org/xdg/xdg-specs/commit/729a8f2dde47694890d87f6e3f91527659f85834

Not sure what would the value be, though.

worldofpeace replied:
How about an optional string `dm.name` upcased that excludes `none`?
```
${optionalString (dm.name != "none") ''
DesktopNames=${dm.name}
''}
```

can't remember if there's a capitalize function in nix.

jtojnar replied:
The purpose of `DesktopNames` is setting `XDG_CURRENT_DESKTOP` environment variable. Given that the variable is primarily used for filtering desktop files, it makes little sense to use our `DE_NAME+WM_NAME`, as nothing will know about them and use them. Well known desktop environments will use upstream session files.

worldofpeace commented on nixos/modules/services/x11/desktop-managers/pantheon.nix:
```diff
@@ -81,9 +81,7 @@ in
-    # If not set manually Pantheon session cannot be started
-    # Known issue of https://github.com/NixOS/nixpkgs/pull/43992
-    services.xserver.desktopManager.default = mkForce "pantheon";
+    services.xserver.displayManager.defaultSession = "pantheon";
```
I'm not sure if this is needed anymore, or if it has any effects on https://github.com/elementary/greeter/issues/368

jtojnar replied:
Hmm, I removed it in one of the reverted commits.

worldofpeace commented:
Per eval definitions of `services.xserver.displayManager.auto` have some missing defaultSession's.

worldofpeace commented on nixos/tests/pantheon.nix:
```diff
@@ -14,6 +14,7 @@ import ./make-test-python.nix ({ pkgs, ...} :
 
     services.xserver.enable = true;
     services.xserver.desktopManager.pantheon.enable = true;
+    services.xserver.displayManager.defaultSession = "pantheon";
```
I don't think this one triggered issues, because the module actually sets it.

worldofpeace commented:
`defaultSession` means which session to select if none has.
It doesn't persist after subsequent selections. This had me confused at first because it looks like it's used for other things. We should adjust the option description.

##### About GDM
What we're doing in GDM is interesting, we set the session for every user unconditionally.
So using `defaultSession` for that actually makes an unexpected behavior.
@jtojnar Should we keep this feature in gdm, but rescope it under a `session` option under `gdm`?
If we're interested in baking in a default-session feature, I could write it into a sub-command in `set-session.py`.

##### About SDDM
It actually [doesn't support AccountsService](https://github.com/sddm/sddm/issues/346) or a setting like [`DefaultSession`](https://github.com/sddm/sddm/issues/326).
The heuristic it has for choosing the "default session" is just whatever session comes first in alphabetical order. So using `defaultSession` here makes no sense because it's not supported.

I'll also quote
> Currently, such a setting exists only for auto-login, but what session we default to for manual login also matters to users.

because this option is implemented just in the autologin configuration
* https://github.com/sddm/sddm/blob/fa5d8436d03c90403ed0a3a397039f88b249d764/data/man/sddm.conf.rst.in#L220

I believe the assert **should** be here, but it should be an option `autoLogin.session`.

It appears opensuse has implemented setting a default session with their alternates mechanism
* https://github.com/sddm/sddm/issues/326#issuecomment-536512758

##### About LightDM
The `user-session` configuration option is scoped per seat, and it's documented as a default.
It actually uses AccountsService so it's already more feature-full.
I believe an assert here **doesn't** make sense.

##### Out of scope thoughts:
Let's make `auto` a module only available for nixos tests.
This module allows root autoLogin, so we would break that for users, but they shouldn't be using it anyways. This gives the impression like `auto` is some special display manager, when it's just lightdm's interface and special pam rules to allow root autoLogin. Its only been useful for NixOS's testing, so I believe this is where it belongs.

worldofpeace commented:
I've added appropriate feature requests in GDM
* https://gitlab.gnome.org/GNOME/gdm/issues/542
* https://gitlab.gnome.org/GNOME/gdm/issues/535

hedning commented:
> defaultSession means which session to select if none has.

Yeah, that's a good point. There's basically two use cases we're interested in:
1. auto login session which should ideally be static
2. first time default session

So, we could introduce another option for auto-login, though I imagine auto-login is what most users care about, and the default session is already mostly used for that. For simplicity I think we could use the defaultSession for both tbh. If so, the description should be updated to make that clear.

It seems like this is what upstream supports:  

<table cellspacing="0" cellpadding="6" rules="groups">


<colgroup>
<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-left">&#xa0;</th>
<th scope="col" class="org-left">GDM</th>
<th scope="col" class="org-left">SDDM</th>
<th scope="col" class="org-left">LightDM</th>
</tr>
</thead>

<tbody>
<tr>
<td class="org-left">auto-login</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">✓</td>
<td class="org-left">✓</td>
</tr>


<tr>
<td class="org-left">first time use</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">&#xa0;</td>
<td class="org-left">✓</td>
</tr>
</tbody>
</table>

For GDM we should probably make the session script only set the session if auto-login is used, or if the session is unset in accounts service (though that's somewhat orthogonal to this PR).

jtojnar commented:
Thanks for the research. I would even say we are squashing together three different options:

|Option|Description|GDM|LightDM|SDDM|
|----|---|---|---|---|
|`initialSession`|Session selected in session chooser when no history is available|➖ [gdm#535](https://gitlab.gnome.org/GNOME/gdm/issues/535)|➕ `user-session` key|➖[sddm#326](https://github.com/sddm/sddm/issues/326)|
|`defaultSession`|Session selected in session chooser at the start of every greeter session (disables session history)|➖ [gdm#542](https://gitlab.gnome.org/GNOME/gdm/issues/542)<br>🤷‍♀️ `./set-session.py`|➖|➖|
|`autologin.session`|Session used for autologin|🤷‍♀️ `./set-session.py`|➕ `autologin-session` key|➕ `Autologin>Session` key|


* `defaultSession` is useful even without auto-login if you want to disable session history – for example you use single DE majority of the time but occasionally you need to use a different one.
* `initialSession` is useless most of the time for DMs with session history, if user switches session!
* The problem with previous implementation was that GDM used `defaultSession` and `defaultSession` had a default value, disabling session history on GDM.

worldofpeace commented:
@jtojnar How you describe `defaultSession` is more like https://gitlab.gnome.org/GNOME/gdm/issues/542, and I think `initial` is just a synonym for `default`.

So I see it more as
* `defaultSession`
* `session`
* `autoLogin.session`

and the latter are implemented and act the same, so I don't think it should even be a separate option.

worldofpeace commented:
@hedning I can add a subcommand to `set-session` to only set the defaultSession if there's no history in AccountsService. I then like being able to force a session without history with a `session` option.
Caring whether if it's for autologin doesn't make sense to me, it should just be force without history or initialize with this value.

hedning commented:
>  12:16 \<jtojnar\> I think we would be fine with just dropping defaultSession altogeher
12:16 \<jtojnar\> and implement autologin session per DM
>  12:24 \<worldofpeace\> So yeah, I think you can drop this weird global default/defaultSession and just have code paths useful for autologin. 

Yeah, this makes sense to me too, the main use-case is auto-login. We're also able to support this for all display managers with a tweak to the setSessionScript which is good :+1:.

I don't think the other use-cases are that important tbh, the most important thing is getting the auto login stuff right, to eg. unblock https://github.com/NixOS/nixpkgs/pull/56988 (looks like we'll get some more bang for deprecating things :).

jtojnar commented:
I wonder if it makes more sense to keep `defaultSession` as an immediately deprecated intermediary step.

hedning commented:
> I wonder if it makes more sense to keep defaultSession as an immediately deprecated intermediary step.

Yeah, using defaultSession as a global proxy should be fine. It fits in with how the default stuff worked before.

worldofpeace commented on nixos/modules/services/x11/display-managers/default.nix:
```diff
@@ -297,18 +250,50 @@ in
+            if cfg.displayManager.defaultSession != null then
+              cfg.displayManager.defaultSession
+            else if cfg.displayManager.sessionData.sessionNames != [] then
+              head installedSessions
```
This is a derivation, not a list.

jtojnar replied:
Away from computer for the next six hours, let's try this:

```suggestion
              head cfg.displayManager.sessionData.sessionNames 
```

jtojnar replied:
Nope, GitHub web UI still does not allow me to commit suggestions, even though I can push to the branch:

> You do not have permission to push to this repository

jtojnar commented:
Rebasing onto master to avoid the conflict.

jtojnar commented:
I think I would like to go with this minimalistic variant for now, running out of free time as the end of the year nears. The per-DM `autologin.session` can be added later.

Ran tests locally, all succeeding except for pantheon which appears to be stuck on https://github.com/NixOS/nixpkgs/blob/aa3fcec6deea9939784a46b555a487e95374193c/nixos/tests/pantheon.nix#L37

@GrahamcOfBorg test gnome3 gnome3-xorg lightdm sddm pantheon plasma5 xfce

worldofpeace commented:
Pantheon test succeeds now @jtojnar.

jtojnar commented:
Well spotted.

Though when I remove `services.xserver.displayManager.defaultSession = "pantheon";` it will still fail:

```
[+16.70s] DEBUG: Greeter requests session default
[+16.71s] DEBUG: Seat seat0: Failed to find session configuration default
[+16.71s] DEBUG: Seat seat0: Can't find session 'default'
```

![Screenshot from 2019-12-12 12-01-25](https://user-images.githubusercontent.com/705123/70706890-27d46a00-1cd7-11ea-8e46-341e9755d0d2.png)

No session will be selected in the greeter:

![Screenshot from 2019-12-12 03-54-36](https://user-images.githubusercontent.com/705123/70706892-27d46a00-1cd7-11ea-97cd-53d9bd49cb4a.png)

This is most likely because LightDM falls back to `user-session=default` and the elementary greeter cannot handle unknown sessions: https://github.com/elementary/greeter/issues/368

This is not an issue with default gtk greeter, which will choose first session in the list when default/last session does not exist.

Also verified that the `user-session` key works as a initialSession/weakDefaultSession.

worldofpeace commented:
> This is most likely because LightDM falls back to user-session=default and the elementary greeter cannot handle unknown sessions: elementary/greeter#368
> 
> This is not an issue with default gtk greeter, which will choose first session in the list when default/last session does not exist.

The second screenshot is weird, I think it should make some sort of selection https://github.com/elementary/greeter/pull/290.

 I may need to find this code in lightdm-gtk-greeter (or comparable in gdm) to cargo cult for elementary/greeter#368 :smile:.

worldofpeace reviewed (changes requested):
Is  `extraSessionFilesPackages` -> `sessionPackages` rename handled?

jtojnar commented:
I confirmed that XMonad works when launched from lightdm (`M-C-x` keyboard launches xterm). But unlike on unstable, the background is missing, need to investigate this.

The tests are broken on unstable as well since the Python port as @worldofpeace noted. The Xmobar also does not seem to do anything so I wonder why it is there.

hedning commented:
I'm seeing a black background in xmonad on this branch and unstable, which seems to be the expected outcome:
https://github.com/NixOS/nixpkgs/blob/856d0a44ec2bb097bd9db6c9363880cf03e06b97/nixos/modules/services/x11/desktop-managers/default.nix#L80

Tested with `~/.background-image` and that worked too.

jtojnar commented:
Yeah, I noticed that too, though on unstable it retains lightdm background for me for some reason.

jtojnar commented:
Rebased onto the sway changes and squashed the commits. Is there anything remaining to do? I successfully ran `gnome-flashback-metacity` from GDM, and `none+xmonad` and `pantheon` from LightDM, and successfully ran all the DE/DM tests.

worldofpeace commented:
It's good with me @jtojnar, and mostly because I don't think I can look over it again differently :smile: 

I'd say, maybe one more reviewer with a different set of eyes.
If waiting doesn't help we can just merge.

hedning commented:
Went over the new code and it looks good to me. :+1: on deprecating the old default options and implementing legacy sessions through sessionPackages. Thanks for bringing this to the finish line :)

worldofpeace commented:
@Profpatsch Agreed to QA this by using it.

grahamc commented:
:heavy_check_mark: I upgraded from 19.09 to nixos-unstable, and added this to my configuration:

```
```

hedning commented:
@grahamc: with this pr LightDM will launch wayland sessions (just double checked that it works with sway), and you'll be able to use `displayManager.defaultSession` to force LightDM to select the sway session by default :)

worldofpeace commented:
> @grahamc: with this pr LightDM will launch wayland sessions (just double checked that it works with sway), and you'll be able to use `displayManager.defaultSession` to force LightDM to select the sway session by default :)

This reminds me I have a branch with https://github.com/canonical/lightdm/commit/03f218981733e50d810767f9d04e42ee156f7feb on top of this. Will PR this after, I'm not sure how it works for sway without the commit though (perhaps it uses the key in the session that forces it)

xaverdh commented:
Just tried this with my `xmonad + lightdm autologin` setup. Works perfectly. I was a bit surprised to get an xterm window though, when not explicitly setting `defaultSession` to "none+xmonad"
(it chose "xterm+xmonad" otherwise) even though I set `services.xserver.desktopManager.xterm.enable` to false. I guess xterm somehow always winds up in the sessions?

xaverdh commented:
Ah sorry `services.xserver.desktopManager.xterm.enable` was *not* set to false (explicitly). The `defaultText` [here](https://github.com/NixOS/nixpkgs/blob/8dc5ff7dcfd1c58c32004ffae25e6d31ed83d86c/nixos/modules/services/x11/desktop-managers/xterm.nix#L18) is misleading. When `stateVersion` is older then "19.09" it defaults to `xSessionEnabled`, not false..

worldofpeace commented:
@xaverdh Autologin requries a defaultSession to be selected. So I believe the it choose `head cfg.displayManager.sessionData.sessionNames` when you didn't set one. I believe the option description describe this well. As for https://github.com/NixOS/nixpkgs/pull/53843#issuecomment-566452499, we hope to eliminate this in 20.03 (default to false) https://github.com/NixOS/nixpkgs/pull/68371.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/setting-up-urxvt-i3/7847/2
